tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -226,16 +232,277 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
       emitNextLine(0)
   }
 
+  // Private API for Rhino
+
+  private[scalajs] object rhinoAPI { // scalastyle:ignore
+    private implicit val GlobalKnowledge: GlobalKnowledge = new Invalidatable {
+      override def invalidate(): Unit = ()
+    }
+
+    def initialize(linkingUnit: LinkingUnit): Unit =
+      startRun(linkingUnit)
+
+    def getHeaderFile(): org.scalajs.core.tools.io.VirtualJSFile =
+      CoreJSLibs.lib(semantics, outputMode)
+
+    def genClassDef(linkedClass: LinkedClass): js.Tree =
+      classEmitter.genClassDef(linkedClass)
+  }
+
+  // Guardian of all global knowledge
+
+  private abstract class InvalidatableBag {
+    def forgetAbout(invalidatable: Invalidatable): Unit
+  }
+
+  private object KnowledgeGuardian { knowledgeGuardian =>
```

Comment:
I don't think this is meaningful/wise. The knowledge guardian is logically coupled to maintaining the caches. Most of `Emitter` is about maintaining caches. There is only a very limited part which deals with the mega-structure of the code emission, and that part is quite directly manipulating the caches, so in fact it is also about the caches.

We could move all of those things together to something like `IncEmitter`, but there wouldn't remain anything in `Emitter` but a call to `incEmitter.emit()` ;-)